### PR TITLE
Proposal: Add isinf()

### PR DESCRIPTION
**Pull request by ichizok**
Currently `isnan()` exists in Vim script, so also adds `isinf()`.

### CODE REVIEW

File: runtime/doc/eval.txt [hunk 1]
```diff
@@ -2252,6 +2252,7 @@ inputsecret({prompt} [, {text}]) String	like input() but hiding the text
 insert({list}, {item} [, {idx}]) List	insert {item} in {list} [before {idx}]
 invert({expr})			Number	bitwise invert
 isdirectory({directory})	Number	|TRUE| if {directory} is a directory
+isinf({expr})			Number	|TRUE| if {expr} is infinity
```
Comment (dpelle) [inline runtime/doc/eval.txt]: How about saying:
```
|TRUE| if Float {expr} is positive or negative infinity
```

File: runtime/doc/eval.txt [hunk 2]
```diff
@@ -5512,6 +5513,13 @@ isdirectory({directory})				*isdirectory()*
 		exist, or isn't a directory, the result is |FALSE|.  {directory}
 		is any expression, which is used as a String.
 
+isinf({expr})						*isinf()*
+		Return |TRUE| if {expr} is a float with an infinity value. >
```
Comment (dpelle) [inline runtime/doc/eval.txt]: Same remark (either positive or negative).

File: src/testdir/test_float_func.vim [hunk 3]
```diff
@@ -288,13 +288,22 @@ func Test_trunc()
   call assert_fails("call trunc('')", 'E808:')
 endfunc
 
+func Test_isinf()
+  call assert_true(isinf(1.0/0.0))
```
Comment (dpelle) [inline src/testdir/test_float_func.vim]: Add test for negative infinity:
```
call assert_true(isinf(-1.0/0.0))
```

File: src/testdir/test_float_func.vim [hunk 4]
```diff
@@ -288,13 +288,22 @@ func Test_trunc()
...
 func Test_isnan()
-  call assert_equal(0, isnan(1.0))
-  call assert_equal(1, isnan(0.0/0.0))
-  call assert_equal(0, isnan(1.0/0.0))
-  call assert_equal(0, isnan('a'))
-  call assert_equal(0, isnan([]))
-  call assert_equal(0, isnan({}))
+  call assert_true(isnan(0.0/0.0))
+  call assert_false(isnan(1.0))
+  call assert_false(isnan(1.0/0.0))
```
Comment (dpelle) [inline src/testdir/test_float_func.vim:304]: Let's also check negative infinity here.

File: src/testdir/test_json.vim [hunk 5]
```diff
@@ -242,7 +242,7 @@ func Test_js_decode()
   call assert_equal(s:varnr, js_decode(s:jsonnr))
   if has('float')
     call assert_equal(s:varfl, js_decode(s:jsonfl))
-    call assert_equal(s:varinf, js_decode(s:jsoninf))
+    call assert_true(isinf(js_decode(s:jsoninf)))
```
Comment (dpelle) [inline src/testdir/test_json.vim]: I just noticed this:
```
:echo js_decode('Infinity')
inf
```
```
:echo js_decode('-Infinity')
E474: Invalid argument
E474: Invalid argument
E474: Invalid argument
v:none
```
Treating js_decode('-Infinity') as an error looks like a bug to me, but it's unrelated to your change.

I've created a separate issue for this: https://github.com/vim/vim/issues/3788

**Comment by brammool** [comment]
Ozaki Kiichi wrote:

> Currently `isnan()` exists in Vim script, so also adds `isinf()`.

The C function isinf() returns -1 for negative infinity, we
should that that as well.

-- 
Drink wet cement and get really stoned.

 /// Bram Moolenaar -- Bram@Moolenaar.net -- http://www.Moolenaar.net   \\\
///        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ \\\
\\\  an exciting new programming language -- http://www.Zimbu.org        ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

**Comment by ichizok** [comment]
Updated document.

@brammool 

> The C function isinf() returns -1 for negative infinity, we
> should that that as well.

C99 (and POSIX) defines that isinf() returns non-zeno if and only if argument is infinity; doesn't distinguish positive and negative.

**Comment by dpelle** [comment]
@ichizok wrote:

> @brammool
> 
> > The C function isinf() returns -1 for negative infinity, we
> > should that that as well.
> 
> C99 (and POSIX) defines that isinf() returns non-zeno if
> and only if argument is infinity; doesn't distinguish positive and negative.

Indeed.  But Linux man page "man isinf", says:
```
      isinf(x)      returns 1 if x is positive infinity, and -1 if x is nega‐
                     tive infinity.
...snip...
NOTES
       In glibc 2.01 and earlier, isinf() returns a nonzero  value  (actually:
       1)  if  x is positive infinity or negative infinity.  (This is all that
       C99 requires.)
```

I'm fine either way.  In any case, user can do something like this
to check for minus infinity:
```
:if isinf(foo) && foo < 0.0
```

**Comment by ichizok** [comment]
> Indeed. But Linux man page "man isinf", says:

Yes, it is about glibc. On other libc (e.g. BSD libc, macOS libc, musl libc) isinf() returns 1 for both positive and negative infinity.

**Comment by brammool** [comment]
> > The C function isinf() returns -1 for negative infinity, we
> > should that that as well.
> 
> C99 (and POSIX) defines that isinf() returns non-zeno if and only if
> argument is infinity; doesn't distinguish positive and negative.

My glibc man page clearly says:

       isinf(x)      returns 1 if x is positive infinity, and -1 if x is nega‐
		     tive infinity.

C99 isn't a very good standard, it has many of these "undefined
behavior" parts in the spec.  I think any sensible person implementing
isinf() would do it like glibc does.  So let's make it that way.

You can use the C isinf() function to check for infinity, then use "val
< 0.0" to check if it's negative.  That way it's portable.


-- 
hundred-and-one symptoms of being an internet addict:
151. You find yourself engaged to someone you've never actually met,
     except through e-mail.

 /// Bram Moolenaar -- Bram@Moolenaar.net -- http://www.Moolenaar.net   \\\
///        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ \\\
\\\  an exciting new programming language -- http://www.Zimbu.org        ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

**Comment by ichizok** [comment]
Hmm, but I think it is odd that the function whose name starts with "is" returns 3 kinds of value (-1/0/1) not boolean-like value.

**Comment by brammool** [comment]
> Hmm, but I think it is odd that the function whose name starts with
> "is" returns 3 kinds of value (-1/0/1) not boolean-like value.

In my opinion following what glibc does is more important.

Otherwise, we would need isinf() (any kind of infinite), isposinf() and
isneginf().  I don't like that, it doesn't have a precedent.

-- 
hundred-and-one symptoms of being an internet addict:
154. You fondle your mouse.

 /// Bram Moolenaar -- Bram@Moolenaar.net -- http://www.Moolenaar.net   \\\
///        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ \\\
\\\  an exciting new programming language -- http://www.Zimbu.org        ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

**Comment by ichizok** [comment]
Updated the patch:
* "isinf()" returns -1/0/1
